Tortuga.Anchor/Tortuga.Anchor/Collections/ListSegment.cs

```diff
@@ -0,0 +1,362 @@
+	/// </summary>
+	private readonly int m_InitialCount;
+
+	/// <summary>
+	/// Initializes a new instance of the <see cref="ListSegment{T}" /> struct.
+	/// </summary>
+	/// <param name="list">The array.</param>
+	/// <exception cref="ArgumentNullException">list</exception>
+	/// <exception cref="ArgumentException">list</exception>
+	public ListSegment(IList<T> list)
+	{
+		if (list == null)
+			throw new ArgumentNullException(nameof(list), $"{nameof(list)} is null.");
+		if (list.Count == 0)
+			throw new ArgumentException($"{nameof(list)} is empty.", nameof(list));
+
+		m_List = list;
+		Offset = 0;
+		Count = list.Count;
+		m_InitialCount = list.Count;
+	}
+
+	/// <summary>
+	/// Initializes a new instance of the <see cref="ListSegment{T}"/> struct.
+	/// </summary>
+	/// <param name="list">The list.</param>
+	/// <param name="offset">The offset.</param>
+	/// <param name="count">The count.</param>
+	/// <exception cref="ArgumentNullException">list</exception>
+	/// <exception cref="ArgumentOutOfRangeException">
+	/// offset
+	/// or
+	/// count
+	/// </exception>
+	public ListSegment(IList<T> list, int offset, int count)
+	{
+		if (list == null)
+			throw new ArgumentNullException(nameof(list), $"{nameof(list)} is null.");
+		if (list.Count == 0)
+			throw new ArgumentException($"{nameof(list)} is empty.", nameof(list));
+
+		if (offset > list.Count)
+			throw new ArgumentOutOfRangeException(nameof(offset), offset, $"{offset} is greater than list size of {list.Count}");
+		if (count > (list.Count - offset))
+			throw new ArgumentOutOfRangeException(nameof(count), count, $"{count} is greater than list size of {list.Count} minus offset of {offset}");
+		if (count == 0)
+			throw new ArgumentOutOfRangeException(nameof(count), count, $"{count} must be greater than zero");
+
+		m_List = list;
+		Offset = offset;
+		Count = count;
+		m_InitialCount = list.Count;
+	}
+
+	/// <summary>
+	/// Gets the count.
+	/// </summary>
+	/// <value>The count.</value>
+	public int Count { get; }
+
+	bool ICollection<T>.IsReadOnly => true;
+
+	/// <summary>
+	/// Gets the underlying list.
+	/// </summary>
+	/// <value>The list.</value>
+	/// <exception cref="InvalidOperationException">The ListSegment is empty.</exception>
+	public IList<T> List
+	{
+		get
+		{
+			CheckSource();
+			return m_List!;
+		}
+	}
+
+	/// <summary>
+	/// Gets the offset.
+	/// </summary>
+	/// <value>The offset.</value>
+	public int Offset { get; }
+
+	/// <summary>
+	/// Gets or sets the value at the specified index.
+	/// </summary>
+	/// <param name="index">The index.</param>
+	/// <returns>T.</returns>
+	/// <exception cref="ArgumentOutOfRangeException">
+	/// index
+	/// or
+	/// index
+	/// </exception>
+	public T this[int index]
+	{
+		get
+		{
+			if (index >= Count)
+				throw new ArgumentOutOfRangeException(nameof(index), index, $"{index} is greater than segment of {Count}");
+
+			return List[Offset + index];
+		}
+		set
```

Comment:
Is ListSegment supposed to be read only?

Reply:
That's code is technically correct. If you cannot add/remove items from a collection, it is read-only even if you can change a value.

    ICollection<int> x = new int[10];
    Console.WriteLine(x.IsReadOnly); //true

That said, we don't need `ListSegment<T>` at this time.